RFC: Allow use of last_error! without extra library calls

ydirson opened the pull request:
Systematic get_last_error() call even when the wrapped call did not return an error looks wasteful.  This change allows to tell the macro that the call has succeeded, and skip the unneeded call.

I have the same feeling towards systematic clear_last_error(): my gut feeling is that this *should* only be useful when calling get_last_error() after the wrapped call succeeds, as on error we're pretty sure `last_error` will be overwritten.  That possibly remains to be double-checked, but situation where it does not hold are likely to be bugs?

ydirson committed:
Allow use of last_error! without extra library calls

Systematic get_last_error() call even when the wrapped call did not
return an error looks wasteful.  This change allows to tell the macro
that the call has succeeded, and skip the unneeded call.

Did not apply to place that require more insights:
- domain_hvm_getcontext: errors are checked for 2nd call only, and then
  only after additional unsafe code


I have the same feeling towards systematic clear_last_error(): my gut
feeling is that this *should* only be useful when calling
get_last_error() after the wrapped call succeeds, as on error we're
pretty sure `last_error` will be overwritten.  That possibly remains
to be double-checked, but situation where it does not hold are likely
to be bugs?

Wenzel commented:
So these are the implementations:

```c
const xc_error *xc_get_last_error(xc_interface *xch)
{
    return &xch->last_error;
}

void xc_clear_last_error(xc_interface *xch)
{
    xch->last_error.code = XC_ERROR_NONE;
    xch->last_error.message[0] = '\0';
}
```

Regardin the suggestion about `xc_clear_last_error()`, I remember that I stumbled on some comments saying that it was a best practise to call this function before any call to `xc_` functions.

I can't find it anymore.
And the xenctrl.h comments doesn't tell much of a story.
I agree that if there is an error, the `xch->last_error` will be overwritten anyway.
And if there isn't, we return the `Ok()` value.

Can you get some guidance from Andrew ?

Wenzel commented:
Can you apply the new `last_error!()` macro to the eligible functions as well ?

ydirson committed:
Remove buggy last_error! use from functions without xc_ calls

In those cases, `get_last_error()` can only return an error from a former
unrelated xc_ call, especially as they don't `clear_last_error()`.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
monitor_enable: remove superfluous last_error! call

This function does call `xc_monitor_enable()` but already checks its
return value and returns in case of error, so `get_last_error()` would
always be called only in case there was no error.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

Wenzel commented:
LGTM
